src/player.ts

```diff
@@ -96,209 +74,54 @@ export class Player {
     }
   }
 
-  private attack() {
-    this.isAttacking = true;
-    this.attackTimer = Date.now();
-  }
-
-  private updateAttack(enemies: Array<Enemy>) {
-    const elapsed = Date.now() - this.attackTimer;
-    if (elapsed >= this.weapon.attackDuration) {
-      this.isAttacking = false;
-    }
-    const weaponRotation =
-      this.facing +
-      this.weapon.swingAngle / 2 -
-      this.weapon.swingAngle *
-        ((Date.now() - this.attackTimer) / this.weapon.attackDuration);
-
-    const weaponCollision = collision.weapon(
-      this.weapon,
-      weaponRotation,
-      this.x,
-      this.y,
-      this.width / 2,
-      enemies
-    );
-    if (weaponCollision.length > 0) {
-      for (let i = 0; i < weaponCollision.length; i++) {
-        enemies.splice(weaponCollision[i], 1);
-      }
+  startAttackTimer() {
+    if (Date.now() - this.attackTimer >= this.weapon.cooldown) {
+      this.isAttacking = true;
+      this.attackTimer = Date.now();
+      setTimeout(() => (this.isAttacking = false), this.weapon.attackDuration);
     }
   }
 
   draw() {
-    if (this.context === null) {
+    if (this.ctx === null) {
       throw new Error('CanvasRenderingContext2D is null.');
     }
-
-    this.context.save();
-    this.context.translate(this.x, this.y);
-    this.context.rotate(this.facing);
-
-    this.context.fillStyle = this.color;
-    this.context.fillRect(
-      (-1 * this.width) / 2,
-      (-1 * this.height) / 2,
-      this.width,
-      this.height
+    this.ctx.save();
+    this.ctx.translate(this.body.x, this.body.y);
+    this.ctx.rotate(this.body.heading);
+
+    this.ctx.fillStyle = this.color;
+    this.ctx.fillRect(
+      (-1 * this.body.width) / 2,
+      (-1 * this.body.height) / 2,
+      this.body.width,
+      this.body.height
     );
 
-    this.context.fillStyle = this.faceColor;
-    this.context.fillRect(
-      (-1 * this.width) / 2,
-      this.height / 2 - 2,
-      this.width,
+    this.ctx.fillStyle = this.faceColor;
+    this.ctx.fillRect(
+      (-1 * this.body.width) / 2,
+      this.body.height / 2 - 2,
+      this.body.width,
       2
     );
 
     if (this.isAttacking) {
       const weaponRotation =
         (this.weapon.swingAngle * (Date.now() - this.attackTimer)) /
         this.weapon.attackDuration;
-      this.weapon.draw(weaponRotation, this.width / 2);
+      this.weapon.draw(weaponRotation, this.body.width / 2);
     }
 
-    this.context.restore();
+    this.ctx.restore();
   }
 
   update(enemies: Array<Enemy>) {
```

Comment:
Doesn't *feel* right to pass enemies through the update function. Maybe you should have some kind of CharacterManager class that you create in game.ts, and pass it into Player's (or even better would be Character's) constructor? CharacterManager can help efficiently retrieve enemies in your vicinity, since you probably don't want to iterate over every enemy everywhere in the game world.

Reply:
I created a characterManager class but differently from what you suggested. Let me know what you think.  82a6944
I haven't decided on my method of filtering the game objects yet. I may make it an issue for later.

Reply:
I think that's useful what you added, but I still think a data structure to hold your enemies/characters would be useful. Fine to handle in a future PR.

Reply:
Check out d9e5b22

Reply:
Future reading suggestion: https://medium.com/@bpmw/quadtrees-for-2d-games-with-moving-elements-63360b08329f